File: 502. IPO.go
func findMaximizedCapital(k int, w int, profits []int, capital []int) int {
	n := len(profits)
	projects := make([]*Project, n)
	for i := range profits {
		projects[i] = newProject(profits[i], capital[i])
	}
	sort.Slice(projects, func(i, j int) bool {
		return projects[i].capital < projects[j].capital
	})
	profitHeap := &MaxHeap{}
	heap.Init(profitHeap)
	for i := 0; i < n && k > 0; i++ {
		if projects[i].capital <= w {
			heap.Push(profitHeap, projects[i])
			continue
		}
		i -= 1
		if profitHeap.Len() == 0 {
			break
		}
		w += heap.Pop(profitHeap).(*Project).profit
		k -= 1
	}
	for k > 0 && profitHeap.Len() > 0 {
		w += heap.Pop(profitHeap).(*Project).profit
		k -= 1
	}
	return w
}

type MaxHeap []*Project

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i].profit > h[j].profit }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MaxHeap) Push(x any) {
	*h = append(*h, x.(*Project))
}

func (h *MaxHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

type Project struct {
	capital int
	profit  int
}

func newProject(profit, capital int) *Project {
	return &Project{profit: profit, capital: capital}
}

File: java/502. IPO.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        PriorityQueue<Project> profitHeap = new PriorityQueue<>((a,b) -> b.profit - a.profit);
        List<Project> projects = new ArrayList<>();
        for (int i = 0; i < profits.length; i++) {
            projects.add(new Project(profits[i], capital[i]));
        }
        Collections.sort(projects, (a,b) -> a.capital - b.capital);
        int answer = w;
        for (int i = 0; i < projects.size() && k > 0; i++) {
            if (projects.get(i).capital <= w) {
                profitHeap.add(projects.get(i));
                continue;
            } else {
                i--;
            }
            if (profitHeap.isEmpty()) break;
            Project current = profitHeap.poll();
            answer += current.profit;
            w += current.profit;
            k--;
        }
        while (!profitHeap.isEmpty() && k > 0) {
            Project current = profitHeap.poll();
            answer += current.profit;
            w -= current.capital;
            w += current.profit;
            k--;
        }
        return answer;
    }
}

class Project {
    int profit;
    int capital;
    Project (int profit, int capital) {
        this.profit = profit;
        this.capital = capital;
    }
}
